From 465a611838247d440a38f4db0adbbbbfce5ded21 Mon Sep 17 00:00:00 2001
From: Subramanian Theinan <stheinan@maxlinear.com>
Date: Thu, 7 Nov 2024 17:48:03 +0800
Subject: [PATCH] WLANRTSYS-83872 [MLD]crashes while giving random mac for AP
 MLD(Negative case) assert @ [SLID:8:68:1:205]

Issue:
When the vendor command to setup mld fails, MLD creation will fail, but without clearing the MLD info
function is returned as failure and since MLD setup is failing the interface are bought down one by one,
since the MLD info was not cleared when MLD setup as failed Destroy MLD is sent to firmware for the vap
which MLD is not present in firmware this is causing firmware crash FW Error Code: [GroupID: 62, FileID: 3, Line: 1267]
as Destroy MLD request is received in invalid MLD state.

Fix:
Clear the MLD info before returning failure when setup MLD vendor command fails.
Related PR : https://mbitbucket.maxlinear.com/projects/SW_WAVE/repos/iwlwav-dev/pull-requests/3842/overview

Fixup to WLANRTSYS-60843 MLO changes: AP MLD configuration.
---
 src/ap/mxl_mld.c | 8 ++------
 1 file changed, 2 insertions(+), 6 deletions(-)

diff --git a/src/ap/mxl_mld.c b/src/ap/mxl_mld.c
index 5b286ca52..933522925 100644
--- a/src/ap/mxl_mld.c
+++ b/src/ap/mxl_mld.c
@@ -336,18 +336,14 @@ static int mxl_setup_link(struct p_ap_mld *mld)
 	/* sending mld_info on link1 context for ADD_MLD */
 	if (MXL_DRV_VENDOR_SET(LTQ_NL80211_VENDOR_SUBCMD_SETUP_MLD, mld_info, sizeof(struct _mxl_vendor_mld_info))) {
 		wpa_printf(MSG_ERROR, "sending vendor mld_info failed !!!");
-		os_free(mld_info);
-		os_free(mld_link_info);
-		return -1;
+		goto err;
 	}
 
 	hapd = link2;
 	/* This will update the ML parameters in link VAP-DB */
 	if (MXL_DRV_VENDOR_SET(LTQ_NL80211_VENDOR_SUBCMD_SET_MLD_INFO, mld_link_info, sizeof(struct _mxl_vendor_mld_link_info))) {
 		wpa_printf(MSG_ERROR, "sending vendor mld_link_info failed on link2!!!");
-		os_free(mld_info);
-		os_free(mld_link_info);
-		return -1;
+		goto err;
 	}
 
 	os_free(mld_info);
-- 
2.43.0

